stonesoup/resampler/particle.py

```diff
@@ -42,7 +43,45 @@ def resample(self, particles):
         u_j = u_i + (1 / n_particles) * np.arange(n_particles)
         index = weight_order[np.searchsorted(cdf, np.log(u_j))]
         new_particles = Particles(state_vector=particles.state_vector[:, index],
-                                  weight=[weight]*n_particles,
+                                  weight=[weight] * n_particles,
                                   parent=Particles(state_vector=particles.state_vector[:, index],
                                                    weight=particles.weight[index]))
         return new_particles
+
+
+class ESSResampler(Resampler):
+    """ESSResampler class
+        This wrapper uses a :class:`~.Resampler` to resample the particles inside
+        an instant of :class:`~.Particles`, but only after checking if this is necessary
+        by comparing Effective Sample Size (ESS) with a supplied threshold (numeric).
+        Kish's ESS is used, as in Doucet-Johansen, 2008:
+            'A Tutorial on Particle Filtering and Smoothing: Fifteen years later'
+        """
+    threshold: float = Property(default=None,
+                                doc='Threshold compared with ESS to decide whether to resample. \
+                                    Default is number of particles divided by 2, \
+                                        set in resample method')
+    resampler: Resampler = Property(default=SystematicResampler,
+                                    doc='Resampler to wrap, which is called \
+                                        when ESS below threshold')
+
+    def resample(self, particles):
+        """
+        Parameters
+        ----------
+        particles : list of :class:`~.Particle`
+            The particles to be resampled according to their weight
+
+        Returns
+        -------
+        particles : list of :class:`~.Particle`
+            The particles, either unchanged or resampled, depending on weight degeneracy
+        """
+        if not isinstance(particles, Particles):
+            particles = Particles(particle_list=particles)
+        if self.threshold is None:
+            self.threshold = len(particles) / 2
```

Comment:
Something to add to the class' `__init__` method, so that this is not needed at every call to `resample`.

Reply:
The problem is, len(particles) is unknown when we create the ESSResampler object.  Typically, the object is created with something like: 
resampler = ESSResampler()
And then later on we have something like:
new_particles = resampler.resample(old_particles)
So the only way I could see to fix this is to maybe make the number of particles a Property in the base Resampler class.

Reply:
Ah got it. That makes sense.